### PR TITLE
Added Fenwick Tree in Java and JavaScript

### LINKED ISSUE #922: Implement Fenwick Tree in Java and JavaScript
I want to add Fenwick Tree code in Two languages one Java and another JavaScript.
@jainaman224 please assign this issue.

## Files to be added
- Fenwick_Tree.java
- Fenwick_Tree.js

### PR DESCRIPTION
**Pull request by dshahid380**
Fixes #922  

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Sample Input / Output.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:

- Adding Fenwick Tree in Java
- Adding Fenwick Tree in JavaScript

#### Languages Used:
- Java
- JavaScript

#### Files Added:
- Fenwick_Tree.java
- Fenwick_Tree.js

### CODE REVIEW

**Comment by dshahid380** [comment]
@SakshiShreya and @jainaman224  please have a look in the screenshots.

#### From line 10 to 41
Screenshot (52)

#### From line 40 to 71
Screenshot (53)

#### From line 73 to 104
Screenshot (54)

#### From line 93 to 121

Screenshot (55)

File: Fenwick_Tree/Fenwick_Tree.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+class FenwickTree {
+    
+    // Fenwick Tree array
+    static int ft[] = new int[10000];
+
+    // Sum operation
+    static int sum(int index) {
+        /*
+	Argument
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 18 is indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        Returns : sum of given arr[0..index]. This function assumes 
+        that the array is preprocessed and partial sums of 
+        array elements are stored in ft[]. 
+        */ 
+        
+        // Initialize sum variable to zero
+	int s = 0;
+	index = index + 1;
+	while(index > 0) {
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Lines 29 to 31 are indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            
+            // Adding tree node to sum
+	    s += ft[index];
+	    index -= index & (-index);
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Lines 34 and 35 are indented with a tab and 4 spaces. Convert to 12 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	return s;
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Lines 36 and 38 are indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    // Update operation
+    static void update(int size, int index, int val) {
+        /*  
+	Arguments
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 44 is indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        index : Index of ft to be updated
+        size  : Length of the original array
+        val   : Add val to the index "index"
+        Traverse all ancestors and add 'val'.
+        Add 'val' to current node of Fenwick Tree. 
+        Update index to that of parent in update.  
+	*/
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 51 is intented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        index : Index of ft to be updated
+        size  : Length of the original array
+        val   : Add val to the index "index"
+        Traverse all ancestors and add 'val'.
+        Add 'val' to current node of Fenwick Tree. 
+        Update index to that of parent in update.  
+	*/ 	
+
+        index += 1;
+	while(index <= size) {
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 54 is indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+	    // Update Fenwick Tree by incrementing index by val
+	    ft[index] += val;
+
+	    // Update parent index
+	    index += index & (-index);
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Lines 56 to 60 are indented with a tab and 4 spaces. Convert to 12 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+
+    // Build Fenwick Tree 
+    static void build(int arr[], int size) {
+        /*
+	Argument
+	arr : The original array
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Lines 67 and 68 are indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        size : The length of the given array
+        This function will construct the Fenwick Tree 
+        from the given array of length "size"
+	*/
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 72 is indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+	arr : The original array
+        size : The length of the given array
+        This function will construct the Fenwick Tree 
+        from the given array of length "size"
+	*/	
+
+        for(int i=0; i<size; i++) {
+
+            // Constructing Fenwick Tree by update operation 
+	    update(size, i, arr[i]);
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 77 is indented with a tab and 4 spaces. Convert to 12 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+	}
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 78 is indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        Print sum of freq[2....7] after update
+        Update position 5 by adding 10
+        Print sum of freq[2....7] after update 
+
+        OUTPUT
+        ------ 
+        12
+        33
+        43 
+	*/
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Line 98 is indented with a tab. Convert to 8 spaces.

File: Fenwick_Tree/Fenwick_Tree.java [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+	update(12, 4, 16);
+       
+        // Print sum fron index = 2 to index = 7
+        System.out.println(sum(7) - sum(2));
+
+        // Increment arr[5] by 10
+	update(12, 5, 10);
+
+        // Print sum from index = 2 to index = 7
+	System.out.println(sum(7) - sum(2));
```
Comment (SakshiShreya) [inline Fenwick_Tree/Fenwick_Tree.java]: Lines 104, 107, 110, 113, 116, 119 are indented with a tab. Convert to 8 spaces.

**Comment by SakshiShreya** [comment]
Also, after looking at the pictures that you sent, I have observed that GitHub uses 8 character tab when you view files. But it uses 6 character tab when you see the commits. (Which is why you are confused that the code looks great but we are finding the mistakes).

And this is a common problem when you use different editors. Some editors use 8 character indentation, some 4 and some 2. That is the reason why we prefer 4 **space** indentation because 1 space takes only 1 character. So, the code will look the same in all editors.

I will suggest you use smart indentation of your editor (if your editor provides it). Otherwise, please convert the tabs into spaces for all the lines that I have mentioned in the review.

And if you have doubts, please feel free to contact us on slack.

**Comment by dshahid380** [comment]
@jainaman224 , @SakshiShreya  and @GOVINDDIXIT please review it. I have made all changes as asked by you.

**Comment by dshahid380** [comment]
Thank you @SakshiShreya, i could not do it without you and @jainaman224

**Pull request merged**